Use test cases from PR for CI

TerrenceMcGuinness-NOAA opened the pull request:
**Description**

This feature changes the CI system to use the test cases from the PR in question. This allows the PR to add or modify tests and have them applied to itself to ensure they work.

Fixes #1662

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**How Has This Been Tested?**

- [ ] Clone and Build tests on WCOSS Dell P3
- [ ] Cycled test on Orion
- [x] Forecast-only test **C48_S2S** on Hera from a case file within the PR

TerrenceMcGuinness-NOAA committed:
added new ff case

TerrenceMcGuinness-NOAA committed:
started PR choises with new element in db and js quere

TerrenceMcGuinness-NOAA committed:
added feature that takes CI-PR-Cases Label from PR and gets cases from the PR reop instead of CI_ROOT

TerrenceMcGuinness-NOAA committed:
removed other cases other than ff for testing

TerrenceMcGuinness-NOAA committed:
couple of updates for debuggin pr cases feature

TerrenceMcGuinness-NOAA committed:
small bug with defining pr_dir

TerrenceMcGuinness-NOAA committed:
two typos forecast and --sbfile

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1662_PRcases

TerrenceMcGuinness-NOAA committed:
fixed bug in driver for getting CI-PR-Cases

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1662_PRcases' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1662_PRcases

TerrenceMcGuinness-NOAA committed:
only C48 cases to test CI-PR-Cases

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1662_PRcases' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1662_PRcases

TerrenceMcGuinness-NOAA committed:
removed case C48_S2SW so to only run C48_SW

TerrenceMcGuinness-NOAA committed:
added back C96C48_hybatmDA C96_atm3DVar after testing cases from PR feature

TerrenceMcGuinness-NOAA committed:
Update driver.sh

double quotes for shell linting

aerorahul reviewed (approved):
looks simple enough

TerrenceMcGuinness-NOAA committed:
Update pr_list_database.py

rando pycodestyle updates

TerrenceMcGuinness-NOAA committed:
changed sbfile to dbfile (eye roll emoji)

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1662_PRcases' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into feature_1662_PRcases

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1662_PRcases

TerrenceMcGuinness-NOAA committed:
added dbfile to database python script

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -86,11 +88,19 @@ for pr in ${pr_list}; do
     done
     rm -Rf "${pr_dir}"
   fi
+  # Check to see if this PR is Labeled CI-PR-Cases to get cases from itself
```
Is there any reason to not just always pull cases from the PR?

TerrenceMcGuinness-NOAA replied:
Not sure I understand the question, the PR is pulled when it is built.

WalterKolczynski-NOAA replied:
Right, but you are relying on whether there is a label set to use the cases from the PR rather than what you already have. Why not always use the ones in the PR? That eliminates the need to set yet another label, and also avoids problems caused by tests being added or changed and the label not being set.

TerrenceMcGuinness-NOAA replied:
Oh, I understand the question now, yes I guess we could just always get cases from the PR, I'm going to refer this question to Raul.

TerrenceMcGuinness-NOAA replied:
The problem I was told is that users can then override the intent of the tests by leaving them out, but like you said, that would be obvious to you.

WalterKolczynski-NOAA commented on ci/scripts/pr_list_database.py:
```diff
@@ -1,12 +1,20 @@
 
 
-def sql_connection(filename: Path) -> sqlite3.Connection:
+def dir_path(string):
```
As dumb as it seems for so simple a function, there should have a doc-block for this.

WalterKolczynski-NOAA replied:
Since this will return the absolute path for a file as well, should it be named `abs_path` of `full_path` instead?

TerrenceMcGuinness-NOAA replied:
This was to make sure it was one or the other and nothing else, I guess you are asking me to over document and split this into two types.

TerrenceMcGuinness-NOAA replied:
Oh wait, I see now, you are asking to simply change the name of the function, yes that makes more sense, will do.

WalterKolczynski-NOAA replied:
Yeah, those were two different things about the same function:
- It needs a doc-block
- If it handles more than just directory paths, the name shouldn't imply it is only for directories

TerrenceMcGuinness-NOAA replied:
- Why does this need a doc block, doesn't that defeat the purpose of useful documentation?
- I did rename it.

WalterKolczynski-NOAA replied:
Because every function/method should have a doc-block. When we finally get around to parsing these to create API documentation, it will be needed. It should be very straightforward and short.

TerrenceMcGuinness-NOAA replied:
ok

TerrenceMcGuinness-NOAA committed:
Update pr_list_database.py

makes sense to call this function name full path not file path because it might not be a file

TerrenceMcGuinness-NOAA reviewed (commented):
updated name of type function for full path (may not be a file)

TerrenceMcGuinness-NOAA committed:
Update driver.sh

Took out the dependency CI-PR-Cases Labels and made getting cases defaulted to the PR's ci/cases directory.

TerrenceMcGuinness-NOAA committed:
f left beheind

TerrenceMcGuinness-NOAA committed:
missing fi in driver

TerrenceMcGuinness-NOAA committed:
added docblock for dir_path and add check for when dbfile arg is not file exept for when create is used

TerrenceMcGuinness-NOAA committed:
added docblock for dir_path and add check for when dbfile arg is not file exept for when create is used

TerrenceMcGuinness-NOAA reviewed (commented):
added doc block and check the file is present except when being created

TerrenceMcGuinness-NOAA reviewed (commented):
added doc block and checks

WalterKolczynski-NOAA commented on ci/scripts/pr_list_database.py:
```diff
@@ -7,7 +7,22 @@
-def dir_path(string):
+def full_path(string):
+    """
+    Type for input argument for dbfile, must be file or path
+
+    Parameters
+    ----------
+    string: string
+        argument sting from --dbfile
+
+    Returns
+    -------
+    os.path.abspath
+        abspath type of input string
```
```suggestion
    Gets the absolute path of the given file and confirms the directory exists

    Parameters
    ----------
    string : str
        Path to a file

    Returns
    --------
    str
        Absolute path of input path
        
    Raises
    -------
    NotADirectoryError
        If the target directory for the file does not exist.
```

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/pr_list_database.py

NICE!

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update pr_list_database.py

white space on line 23
